### PR TITLE
Fix `@JsonAnyGetter` keys not prefixed when inner bean uses `@JsonUnwrapped(prefix=...)`

### PR DESCRIPTION
**Pull request by pjfanning**
#6122 
#6118

When `@JsonAnyGetter` appears on a bean referenced via `@JsonUnwrapped(prefix="...")`, the prefix was never applied to the any-getter map keys. Worse, `AnyGetterWriter` didn't override `rename(NameTransformer)`, so the inherited path called `_new(PropertyName)` which throws `IllegalStateException` for subclasses — crashing serializer construction.

## Example

```java
class Outer {
    public String name = "aaa";
    @JsonUnwrapped(prefix = "a-")
    public Inner inner; // inner.getExtra() returns {"age": "64"}
}
// Before fix: IllegalStateException at mapper construction, or wrong output
// After fix:  {"name":"aaa","a-age":"64"}
```

### `AnyGetterWriter`
- Added `NameTransformer _nameTransformer` field (null in normal, non-unwrapped use)
- Added protected copy constructor `AnyGetterWriter(AnyGetterWriter src, NameTransformer transformer)` with correct chaining: `chainedTransformer(newTransformer, src._nameTransformer)`, matching the ordering in `UnwrappingBeanPropertyWriter`
- Overrode `rename(NameTransformer)` to return a copy storing the transformer rather than attempting to rename the property itself (meaningless for any-getter)
- `getAndSerialize()` and `getAndFilter()` now dispatch to transformer-aware helpers when `_nameTransformer != null`, applying it to each map/`ObjectNode` key individually

### `UnwrappedWithAnySetterTest`
- Added `testUnwrappedWithPrefixWithAnyGetter` covering the bug scenario: outer direct field serializes without prefix; inner any-getter entries get the `"a-"` prefix applied per key

### CODE REVIEW

**Comment by pjfanning** [comment]
There is still an issue with deserialization of maps when unwrapped with prefix is used. The deserialization of the map doesn't remove the prefix. I want to treat that as a separate issue and can work out a fix for that if and when this serialization fix is merged.

**Comment by pjfanning** [comment]
@cowtowncoder I fixed the deserialization issue with AI assistance. The issue I described above (https://github.com/FasterXML/jackson-databind/pull/6123#issuecomment-5083230805)

File: src/main/java/tools/jackson/databind/deser/bean/BeanDeserializerBase.java [hunk 1]
```diff
@@ -194,6 +194,16 @@ public abstract class BeanDeserializerBase
      */
     protected UnwrappedPropertyHandler _unwrappedPropertyHandler;
 
+    /**
+     * When this deserializer was created via {@code unwrappingDeserializer()},
+     * this holds the {@link NameTransformer} so that unknown property names
+     * can be reverse-transformed before being passed to
+     * {@code @JsonAnySetter}.
+     *
+     * @since 3.3
```
Comment (pjfanning) [inline src/main/java/tools/jackson/databind/deser/bean/BeanDeserializerBase.java:203]: I have put version 3.3 on all the since markers but can change them to 3.1 or 3.2 if this is deemed safe for backporting

**Comment by cowtowncoder** [comment]
Ok. AI review pointed out a seemingly serious issue:

-----

### Correctness issue — confirmed regression (high severity)

AnyGetterWriter._serializeMapEntriesWithTransformer() (the new fallback used whenever _nameTransformer != null) **completely bypasses _mapSerializer for real java.util.Map values, replacing it with a hand-rolled loop: rawKey.toString() for keys and ctxt.writeValue(gen, entryValue) for values**. This silently drops everything MapSerializer normally provides:
  
- Content inclusion filtering (@JsonInclude(NON_NULL)/NON_EMPTY etc. on the any-getter map) — I reproduced this concretely: with @JsonInclude(NON_NULL) on the extra map, a null-valued entry that's correctly suppressed when the bean is not unwrapped is emitted ("a-nickname":null) once the same bean is referenced via `@JsonUnwrapped(prefix="a-")`.
- Custom key serializers (_keySerializer, e.g. @JsonSerialize(keyUsing=...), non-String keys with custom formatting) — silently downgraded to Object.toString().
 - Whatever else MapSerializer.serializeFilteredAnyProperties()/serializeWithoutTypeInfo() do beyond the naive loop (polymorphic value typing, etc.).

This means: for any bean that combines `@JsonAnyGetter` + `@JsonUnwrapped(prefix/suffix=...)`, Map serialization semantics silently regress the moment a name-transform is in play — trading one bug (missing prefix) for a different, arguably worse one (silently wrong output for filtered/keyed maps). A better approach would push the key transformation into the existing MapSerializer path (e.g., wrap/rename the key after _keySerializer runs, or pass the transformer into  serializeWithoutTypeInfo/serializeFilteredAnyProperties) rather than routing around it.

------

I'll see if that can be fixed; that cannot be merged as-is.

**Comment by pjfanning** [comment]
@cowtowncoder I got AI to sort some of the issues but there might be more.

**Comment by cowtowncoder** [comment]
@pjfanning Hoping to pick this up soon, maybe tomorrow.

**Comment by pjfanning** [comment]
I got Claude Opus to write extra tests and it has 7 that fail and they are annotated so they don't run.

# Test coverage for `@JsonUnwrapped(prefix)` + `@JsonAnyGetter`/`@JsonAnySetter`

Added 24 tests across two files. All pass; the 7 in `tofix/` pass *because* they fail — each is a real
edge case where the PR ([#6123](https://github.com/FasterXML/jackson-databind/pull/6123)) breaks down.

## What works — `struct/UnwrappedWithAnyGetterPrefixTest.java` (17 tests)

- prefix / suffix / prefix+suffix round-trips
- plain `@JsonUnwrapped` (NOP transformer) unchanged
- keys that already contain the prefix (`a-age` → `a-a-age` → `a-age`)
- empty any-getter map writes nothing
- any-getter/any-setter declared on a **field**
- `ObjectNode`-valued any-getter (`[databind#3604]`) — nested node keys correctly left alone
- **enum** map keys
- custom (non-`MapSerializer`) any-getter serializer via `@JsonSerialize(using=...)`
- property-based **creator** on the unwrapped bean
- `@JsonIdentityInfo` outer, polymorphic (`@JsonTypeInfo`) outer
- nested `@JsonUnwrapped` prefixes chaining on **write** (`a-b-age`)
- two prefixed unwrapped beans on **write**
- handler-cache safety: same inner type used with different prefixes *and* standalone

## What doesn't — `tofix/UnwrappedAnyGetterPrefixTest.java` (7 tests)

| # | Failure | Cause |
|---|---------|-------|
| 1 | `{"a-obj":{"a-x":1,"a-y":2}}` — nested Map/Array keys prefixed | `NameTransformingGenerator` intercepts *every* `writeName`, at any depth, not just the any-getter's own keys |
| 2 | `{"a-pt":{"a-x":1,"a-y":2}}` — nested POJO properties prefixed | same |
| 3 | `{"a-pet":{"a-@type":"dog",…}}` — polymorphic **type id** prefixed | same |
| 4 | `{"3":"x"}` — `Integer`/`Long` keys get no prefix | those go through `JsonGenerator.writePropertyId(long)`, which the delegate doesn't override (String/enum/`byte[]` keys use `writeName` and do work) |
| 5 | Builder-based inner: any-setter receives `a-age` | `BuilderBasedDeserializer.unwrappingDeserializer()` doesn't retain the `NameTransformer` the way `BeanDeserializer` now does |
| 6 | Two prefixed unwrapped beans: each any-setter collects the other's props (`a.ea = {p=1, b-q=2}`) | when `reverse()` returns `null` the raw name is passed through instead of being skipped — writes are correct, so the value doesn't round-trip |
| 7 | Nested `@JsonUnwrapped` prefixes: `a-b-age` silently dropped on read | pre-existing, but write now emits `a-b-age`, so the new serialization side has no working read side |

**Provenance:** items 1–4 are new with this branch — on the base commit (`30e603d3f`) this whole scenario
threw `IllegalStateException: Method must be overridden by class AnyGetterWriter`. Items 5 and 6 are
unchanged-but-wrong. Item 7 predates the branch (verified in a worktree at `30e603d3f`).
Items 1–3 were checked against a non-unwrapped baseline, which correctly produces `{"obj":{"x":1}}`.

## Suggested fixes

1. **1–3 (the serious one)** — a generator wrapper can't tell depth apart. Either make
   `NameTransformingGenerator` transform only at the depth it was installed (track
   `writeStartObject`/`writeEndObject` nesting, transform only when depth == 0), or drop the wrapper
   and transform keys explicitly in a `MapSerializer` entry loop.
2. **4** — override `writePropertyId(long)` in `NameTransformingGenerator`:
   `delegate.writeName(_transformer.transform(String.valueOf(id)))`.
3. **5** — mirror the `BeanDeserializer` change in `BuilderBasedDeserializer`.
4. **6** — skip the property (rather than passing the raw name) when `reverse()` returns `null`.
   Safe, since the no-prefix case uses `NameTransformer.NOP`, whose `reverse()` never returns `null`.

File: src/main/java/tools/jackson/databind/deser/bean/BeanDeserializerBase.java [hunk 2]
```diff
@@ -1983,6 +1998,45 @@ protected boolean _shouldSkipUnknowns(DeserializationContext ctxt) {
...
+                return reversed;
+            }
+        }
+        return propName;
+    }
+
+    /**
+     * [databind#6118] Companion to {@link #_anySetterKey}: tells whether the
+     * {@code @JsonAnySetter} of this bean (if any) should be given the property at all.
+     *<p>
```
[the remark this answers is not in the document]
Comment (pjfanning) [reply src/main/java/tools/jackson/databind/deser/bean/BeanDeserializerBase.java]: fixed
